module/state_synchronization/requester/jobs/execution_data_reader.go

```diff
@@ -0,0 +1,130 @@
+package jobs
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/state_synchronization"
+	"github.com/onflow/flow-go/module/util"
+	"github.com/onflow/flow-go/storage"
+)
+
+// BlockEntry represents a block that's tracked by the ExecutionDataRequester
+type BlockEntry struct {
+	BlockID       flow.Identifier
+	Height        uint64
+	ExecutionData *state_synchronization.ExecutionData
+}
+
+// ExecutionDataReader provides an abstraction for consumers to read blocks as job.
+type ExecutionDataReader struct {
+	component.Component
+	cm *component.ComponentManager
+
+	head func() uint64
+	read ReadExecutionData
+
+	maxCachedEntries uint64
+	cache            map[uint64]*state_synchronization.ExecutionData
+
+	// TODO: refactor this to accept a context in AtIndex instead of storing it on the struct.
+	// This requires also refactoring jobqueue.Consumer
+	ctx irrecoverable.SignalerContext
+
+	mu sync.RWMutex
+}
+
+type ReadExecutionData func(irrecoverable.SignalerContext, uint64) (*state_synchronization.ExecutionData, error)
+
+// NewExecutionDataReader creates and returns a ExecutionDataReader.
+func NewExecutionDataReader(maxCachedEntries uint64, fetchHead func() uint64, fetchData ReadExecutionData) *ExecutionDataReader {
```

Comment:
That method implements the `Jobs.Head()` method which is why I named it that. I can pick a new name. It's not actually related to the caching though. It's just used for that since it's simpler to cache based on head than the last processed index of the consumer that uses the reader